scripts/normalize-by-median.py

```diff
@@ -55,166 +48,111 @@ def WithDiagnostics(ifile, batch_size, fp, paired, norm):
 
     index = 0
 
-    for index, batch in enumerate(batchwise(
-                                  screed.open(ifile, parse_description=False),
-                                  batch_size)):
+    fp = norm.report_fp
 
-        norm.total += batch_size
-        total = norm.total
-        discarded = norm.discarded
+    # per read diagnostic output
+    for index, is_paired, read0, read1 in reader:
 
         if index > 0 and index % 100000 == 0:
             print('... kept {kept} of {total} or {perc:2}%'
-                  .format(kept=total - discarded,
-                          total=total,
-                          perc=int(100. - discarded / float(total) * 100.)),
+                  .format(kept=norm.total - norm.discarded,
+                          total=norm.total,
+                          perc=int(100. - norm.discarded /
+                                   float(norm.total) * 100.)),
                   file=sys.stderr)
 
-            print('... in file ' + input_filename, file=sys.stderr)
+            print('... in file ' + ifilename, file=sys.stderr)
 
-            if report_fp:
+            if fp:
                 print(total + " " + total - discarded + " " +
                       1. - (discarded / float(total)), file=fp)
                 report_fp.flush()
 
-        # If in paired mode, check that the reads are properly interleaved
-        if paired:
-            if not check_is_pair(batch[0], batch[1]):
-                raise IOError('Error: Improperly interleaved pairs \
-                    {b0} {b1}'.format(b0=batch[0].name, b1=batch[1].name))
+        for record in norm(index, is_paired, read0, read1):
```

Comment:
On Fri, Jun 05, 2015 at 12:56:10PM -0700, Jake Fenton wrote:
> >                  print(total + " " + total - discarded + " " +
> >                        1. - (discarded / float(total)), file=fp)
> >                  report_fp.flush()
> >  
> > -        # If in paired mode, check that the reads are properly interleaved
> > -        if paired:
> > -            if not check_is_pair(batch[0], batch[1]):
> > -                raise IOError('Error: Improperly interleaved pairs \
> > -                    {b0} {b1}'.format(b0=batch[0].name, b1=batch[1].name))
> > +        for record in norm(index, is_paired, read0, read1):
> 
> Wouldn't that muck up how the generators work?

nope :)

just have the for loop in WithDiagnostics loop over norm(reader).